File: example/src/main/java/net/dontdrinkandroot/wicket/example/page/NavBarPage.java
package net.dontdrinkandroot.wicket.example.page;

import org.apache.wicket.Component;
import org.apache.wicket.markup.html.WebMarkupContainer;
import org.apache.wicket.markup.html.link.AbstractLink;
import org.apache.wicket.markup.html.link.BookmarkablePageLink;
import org.apache.wicket.markup.repeater.RepeatingView;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.Model;
import org.apache.wicket.request.mapper.parameter.PageParameters;

import net.dontdrinkandroot.wicket.bootstrap.component.item.BookmarkablePageLinkItem;
import net.dontdrinkandroot.wicket.bootstrap.component.item.DividerItem;
import net.dontdrinkandroot.wicket.bootstrap.component.item.DropDownItem;
import net.dontdrinkandroot.wicket.bootstrap.component.item.HeaderItem;
import net.dontdrinkandroot.wicket.bootstrap.component.item.NavBarButtonItem;
import net.dontdrinkandroot.wicket.bootstrap.component.item.NavBarTextItem;
import net.dontdrinkandroot.wicket.bootstrap.component.navbar.InnerNavBar;
import net.dontdrinkandroot.wicket.bootstrap.css.NavBarStyle;


public class NavBarPage extends ComponentPage
{

	public NavBarPage(PageParameters parameters)
	{
		super(parameters);
	}

	@Override
	protected IModel<String> createPageHeadingModel()
	{
		return Model.of("Nav Bars");
	}

	@Override
	protected void onInitialize()
	{
		super.onInitialize();

		InnerNavBar navBarDefault = new InnerNavBar("navBarDefault") {

			@Override
			protected Component createBrandLink(String id)
			{
				BookmarkablePageLink<Void> brandLink = new BookmarkablePageLink<Void>(id, HomePage.class);
				brandLink.setBody(Model.of("Brand"));
				return brandLink;
			}

			@Override
			protected Component createNavBarForm(String id)
			{
				return NavBarPage.this.createExampleNavBarForm(id);
			}

			@Override
			protected void populateNavbarLeftItems(RepeatingView navbarLeftItemView)
			{
				NavBarPage.this.populateExampleNavbarLeftItems(navbarLeftItemView);
			}

			@Override
			protected void populateNavbarRightItems(RepeatingView navbarRightItemView)
			{
				NavBarPage.this.populateExampleNavbarRightItems(navbarRightItemView);
			}
		};
		this.add(navBarDefault);

		InnerNavBar navBarInverse = new InnerNavBar("navBarInverse") {

			@Override
			protected Component createBrandLink(String id)
			{
				BookmarkablePageLink<Void> brandLink = new BookmarkablePageLink<Void>(id, HomePage.class);
				brandLink.setBody(Model.of("Brand"));
				return brandLink;
			}

			@Override
			protected Component createNavBarForm(String id)
			{
				return NavBarPage.this.createExampleNavBarForm(id);
			}

			@Override
			protected void populateNavbarLeftItems(RepeatingView navbarLeftItemView)
			{
				NavBarPage.this.populateExampleNavbarLeftItems(navbarLeftItemView);
			}

			@Override
			protected void populateNavbarRightItems(RepeatingView navbarRightItemView)
			{
				NavBarPage.this.populateExampleNavbarRightItems(navbarRightItemView);
			}
		};
		navBarInverse.setStyle(NavBarStyle.INVERSE);
		this.add(navBarInverse);
	}

	protected Component createExampleNavBarForm(String id)
	{
		return new WebMarkupContainer(id);
	}

	protected void populateExampleNavbarLeftItems(RepeatingView itemView)
	{
		itemView.add(new DropDownItem(itemView.newChildId(), Model.of("DropDown")) {

			@Override
			protected void populateItems(RepeatingView itemView)
			{
				itemView.add(new BookmarkablePageLinkItem(itemView.newChildId(), Model.of("Action"), HomePage.class));
				itemView.add(new DividerItem(itemView.newChildId()));
				itemView.add(new HeaderItem(itemView.newChildId(), Model.of("A Header")));
				itemView.add(
						new BookmarkablePageLinkItem(
								itemView.newChildId(),
								Model.of("Another Action"),
								HomePage.class));

			}
		});
		itemView.add(new NavBarButtonItem<Void>(itemView.newChildId()) {

			@Override
			protected AbstractLink createLink(String id)
			{
				return new BookmarkablePageLink<Void>(id, HomePage.class).setBody(Model.of("Button"));
			}
		});

	}

	protected void populateExampleNavbarRightItems(RepeatingView itemView)
	{
		itemView.add(new NavBarTextItem(itemView.newChildId(), Model.of("Text")));
	}
}

File: example/src/main/kotlin/net/dontdrinkandroot/wicket/example/page/component/NavbarPage.kt
package net.dontdrinkandroot.wicket.example.page.component

import net.dontdrinkandroot.wicket.behavior.CssClassAppender
import net.dontdrinkandroot.wicket.bootstrap.component.item.divider
import net.dontdrinkandroot.wicket.bootstrap.component.item.dropdown
import net.dontdrinkandroot.wicket.bootstrap.component.item.header
import net.dontdrinkandroot.wicket.bootstrap.component.item.pageLink
import net.dontdrinkandroot.wicket.bootstrap.component.navbar.*
import net.dontdrinkandroot.wicket.bootstrap.css.BackgroundColor
import net.dontdrinkandroot.wicket.bootstrap.css.NavbarAlignment
import net.dontdrinkandroot.wicket.bootstrap.css.NavbarStyle
import net.dontdrinkandroot.wicket.example.component.NavbarForm
import net.dontdrinkandroot.wicket.example.page.HomePage
import net.dontdrinkandroot.wicket.markup.html.link.BookmarkablePageLink
import net.dontdrinkandroot.wicket.model.model
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model
import org.apache.wicket.request.mapper.parameter.PageParameters

class NavbarPage(parameters: PageParameters) : ComponentPage(parameters) {

    override fun createPageHeadingModel() = "Navbars".model()

    override fun onInitialize() {
        super.onInitialize()
        val navbarLight = createExampleNavbar("navbarLight")
        navbarLight.add(CssClassAppender(BackgroundColor.LIGHT))
        this.add(navbarLight)

        val navbarDark = createExampleNavbar("navbarDark", styleModel = Model(NavbarStyle.DARK))
        navbarDark.add(CssClassAppender(BackgroundColor.DARK))
        this.add(navbarDark)
    }

    protected fun createExampleNavbar(id: String, styleModel: IModel<NavbarStyle> = Model(null)): Navbar = navbar(
        id,
        styleModel = styleModel,
        createBrandHandler = { id ->
            BookmarkablePageLink<Void>(id, pageClass = HomePage::class.java, bodyModel = Model("Brand"))
        }
    )
    {
        navbarNav()
        {
            dropdown("Dropdown") {
                pageLink("Action", HomePage::class)
                divider()
                header("A header")
                pageLink("Another Action", HomePage::class)
            }
            pageLink("Link", NavbarPage::class)
        }

        val form = NavbarForm(newChildId())
        add(form)

        val text = NavbarText(newChildId(), Model.of("Text"))
        add(text)
        add(NavbarButtonLink<Void>(
            newChildId(),
            bodyModel = Model("Button"),
            alignmentModel = NavbarAlignment.RIGHT.model()
        ) {})
    }
}